### PR TITLE
fix: preserve dynamic entry functions (#48)

### PR DESCRIPTION
**Pull request by crlfe**
Maintainer has objections on the issue, so am tossing this up to make sure we're talking about the same thing.

This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Dynamic entry functions do not work because they are discarded when sanitizing the configuration.

### Breaking Changes

No

### Additional Info

Unrolled the ['entry', 'output'] loop to keep exactly the same behavior on the result.output object.

### CODE REVIEW

File: lib/compiler.js [hunk 1]
```diff
@@ -76,12 +76,16 @@ function sanitize(config) {
...
-    for (const property of ['entry', 'output']) {
-      const target = result[property];
-      if (target && Object.keys(target).length === 0) {
-        delete result[property];
-      }
+    // TODO: remove the need for these cleanups
+    if (
+      result.entry &&
+      Object.keys(result.entry).length === 0 &&
+      !(result.entry instanceof Function)
```
Comment (shellscape) [inline lib/compiler.js]: The `instanceof` check here is flawed:

``` console
$ const p = new Promise((r, x)=>{})
$ p instanceof Function
false
$ typeof p
'object'
$ typeof p.then
'function'
$ const f = () => {}
$ typeof f
'function'
```

**Review by shellscape — CHANGES_REQUESTED** [review]

File: lib/compiler.js (hunk 1 above)
Comment (crlfe) [reply lib/compiler.js]: Neither the docs nor webpack allow entry to be PromiseLike (and changing that would be a breaking for anyone with a 'then' entrypoint).

The case you're likely thinking of is that entry may be a function that _returns_ PromiseLike.

**Comment by crlfe** [comment]
For the codecov check, would you rather bypass that check for this PR or have me add some mostly-useless tests of empty entry and output?

If the latter, propose I create fixtures/cli/ for bad-config, zero-config and the new function-entry-config, empty-entry-config, and empty-output-config. (Thoughts on removing the trailing `-config`?)

**Comment by crlfe** [comment]
Abandoning, since I'm no longer using these tools, and prefer not to waste my time.
